File: scraper/sources/tm.py
"""
TM — Telekom Malaysia Tender Notices.
URL: https://www.tm.com.my/business-with-tm/tender-notices

Playwright required.
"""
import logging
from typing import Iterator

from scraper.utils import parse_date, infer_status, now_iso

SOURCE_ID = 12
SOURCE_NAME = "TM"
BASE_URL = "https://www.tm.com.my/business-with-tm/tender-notices"
HOST = "https://www.tm.com.my"

logger = logging.getLogger(__name__)


def scrape() -> Iterator[dict]:
    try:
        from playwright.sync_api import sync_playwright
    except ImportError:
        logger.warning("Playwright not installed — skipping %s", SOURCE_NAME)
        return

    with sync_playwright() as pw:
        browser = pw.chromium.launch(headless=True)
        page = browser.new_page()
        try:
            page.goto(BASE_URL, timeout=30000, wait_until="networkidle")
            page.wait_for_selector("table tbody tr, .tender-list li, article", timeout=15000)

            for row in page.query_selector_all("table tbody tr, .tender-list li"):
                cells = [c.inner_text().strip() for c in row.query_selector_all("td, span")]
                if len(cells) < 2:
                    continue
                link = row.query_selector("a")
                url = link.get_attribute("href") if link else BASE_URL
                if url and url.startswith("/"):
                    url = HOST + url

                deadline = parse_date(cells[2]) if len(cells) > 2 else None
                open_d   = parse_date(cells[3]) if len(cells) > 3 else None
                now = now_iso()
                yield {
                    "source_id": SOURCE_ID,
                    "ref":       cells[0],
                    "title":     cells[1],
                    "deadline":  deadline,
                    "open_date": open_d,
                    "category":  "Telecommunications",
                    "ministry":  "Telekom Malaysia Berhad",
                    "status":    infer_status(open_d, deadline),
                    "url":       url or BASE_URL,
                    "scraped_at": now,
                }
        except Exception as exc:
            logger.error("%s scrape error: %s", SOURCE_NAME, exc)
        finally:
            browser.close()

File: scraper/sources/tm.js
'use strict';

// TM (Telekom Malaysia) — React SPA, uses Playwright to render tender notices page
// URL: https://www.tm.com.my/business-with-tm/tender-notices
const { chromium } = require('playwright');
const { parseDate, inferStatus, nowIso } = require('../utils');

const SOURCE_ID = 12;
const SOURCE_NAME = 'TM';
const BASE_URL = 'https://www.tm.com.my/business-with-tm/tender-notices';
const HOST = 'https://www.tm.com.my';

async function* scrape() {
  const now = nowIso();
  let totalYielded = 0;
  let browser = null;

  try {
    browser = await chromium.launch({
      headless: true,
      args: ['--no-sandbox', '--disable-setuid-sandbox', '--disable-dev-shm-usage'],
    });
    const ctx = await browser.newContext({
      userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36',
      locale: 'en-MY',
      viewport: { width: 1366, height: 768 },
    });
    const page = await ctx.newPage();
    await page.route('**/*', (route) => {
      if (['image', 'media', 'font'].includes(route.request().resourceType())) route.abort();
      else route.continue();
    });

    await page.goto(BASE_URL, { waitUntil: 'networkidle', timeout: 60000 });
    await page.waitForTimeout(3000);

    // Wait for tender content to appear
    try {
      await page.waitForSelector('table tr td, .tender-list, article, [class*="tender"]', { timeout: 20000 });
    } catch (_) {}

    const html = await page.content();
    const cheerio = require('cheerio');
    const $ = cheerio.load(html);
    const items = [];

    // Table approach
    $('table tr').each((_, tr) => {
      const cells = $(tr).find('td').map((_, td) => $(td).text().trim()).get();
      if (cells.length < 2) return;
      const title = cells[1] || cells[0];
      if (!title || title.length < 15) return;
      const dates = cells.filter(c => /\d{1,2}[\/\-]\d{1,2}[\/\-]\d{4}/.test(c));
      const closeRaw = dates[dates.length - 1] || null;
      const link = $(tr).find('a').attr('href');
      const url  = link ? new URL(link, BASE_URL).href : BASE_URL;
      items.push({ title, closeRaw, url });
    });

    // Card / list approach
    $('article, [class*="tender"], [class*="notice"], li').each((_, el) => {
      const title = $(el).find('h2,h3,h4,a,.title').first().text().trim();
      if (!title || title.length < 15) return;
      const text = $(el).text();
      const dates = text.match(/\d{1,2}[\/\-]\d{1,2}[\/\-]\d{4}/g) || [];
      const closeRaw = dates[dates.length - 1] || null;
      const link = $(el).find('a').attr('href');
      const url  = link ? new URL(link, BASE_URL).href : BASE_URL;
      items.push({ title, closeRaw, url });
    });

    console.log(`[${SOURCE_NAME}] ${items.length} items`);
    for (const item of items) {
      const deadline = parseDate(item.closeRaw);
      yield {
        source_id: SOURCE_ID, ref: null,
        title: item.title, category: null, ministry: 'TM',
        open_date: null, deadline, status: inferStatus(null, deadline),
        url: item.url, scraped_at: now,
      };
      totalYielded++;
    }

    console.log(`[${SOURCE_NAME}] done — ${totalYielded} records`);
  } catch (err) {
    console.error(`[${SOURCE_NAME}] fatal: ${err.message}`);
  } finally {
    try { if (browser) await browser.close(); } catch (_) {}
  }
}

module.exports = { SOURCE_ID, SOURCE_NAME, scrape };
